Fix folder size contained in S3 buckets

artonge opened the pull request:
Folders contained in S3 bucket's are displayed with a size of "Pending".

The following command can get the real size to be displayed, but only for one refresh: `php occ files:scan --path "<S3 bucket path>"`

Rational (I might be wrong):
1. When running the `stat` method on the S3 folder, the `mtime` is always set to 'now'.
2. This leads to the `hasUpdated` method from `Files/Storage/Common.php` to always returns `true`,
3. And to the `needsUpdate` method in `Files/View.php` to always returns `true` too.
4. So the cache gets updated by `Files/Cache/Scanner.php` every time the folder is acceceed. But it gets updated with a size of `-1` because the size for S3 folders is not getting computed and default to `-1`

Need guidance on how and where we should address this issue :).

juliushaertl commented on lib/private/Files/Cache/Scanner.php:
```diff
@@ -396,6 +396,8 @@ protected function scanChildren($path, $recursive = self::SCAN_RECURSIVE, $reuse
 		}
 		if ($this->cacheActive) {
+			// 4. The cached to be updated every time the folder is acceceed.
+			// But this is always called with -1
```
Did you check the code path on where the -1 is coming from? Wonders me a bit that the occ file scan produces the proper size but not the scan on cache update

artonge replied:
I tried, but got lost in the layers :/, it could come from a lot of places. I think the difference between file scan and UI access is this condition: https://github.com/nextcloud/server/blob/e05aa2cd25013336a0201c3a0a5396f3c609a49c/lib/private/Files/Cache/Scanner.php#L436

juliushaertl commented on apps/files_external/lib/Lib/Storage/AmazonS3.php:
```diff
@@ -386,6 +386,7 @@ public function stat($path) {
 			if ($this->is_dir($path)) {
 				//folders don't really exist
 				$stat['size'] = -1; //unknown
+				// 1. The fact that mtime is always 'now' leads to...
```
Not fully sure but since https://github.com/nextcloud/server/blob/215aef3cbdc1963be1bb6bca5218ee0a4b7f1665/lib/private/Files/Cache/Watcher.php#L127 checks for the storage mtime, it might be enough to update that one instead of the actual mtime to trigger the cache update. Or we could ensure that the scan also triggers the cache update if only the size changed to -1. 

I guess this is only an issue if the external storage is configured to always check for changes? As the other case should have been covered by https://github.com/nextcloud/server/pull/21530 / https://github.com/nextcloud/server/pull/21728

artonge replied:
Changing `storage_mtime` instead of `mtime` works, but I am not sure about the side effects and implications of that kind of change.

> I guess this is only an issue if the external storage is configured to always check for changes? As the other case should have been covered by #21530 / #21728

This is correct. It only fails when configured to always check.

artonge replied:
Changing storage_mtime instead of mtime works, but I am not sure about the side effects and implications of that kind of change.

> I guess this is only an issue if the external storage is configured to always check for changes? As the other case should have been covered by #21530 / #21728

This is correct. It only fails when configured to always check.

PVince81 replied:
you might need to be careful: the data that comes back from the storage implementation might not match 100% what's in the cache. Basically, there **might** be a place where $data['mtime']  gets converted to $cacheData['storage_mtime']. better double check that

artonge commented on apps/files_external/lib/Lib/Storage/AmazonS3.php:
```diff
@@ -384,13 +384,14 @@ public function stat($path) {
-				$stat['mtime'] = time();
 				$cacheEntry = $this->getCache()->get($path);
-				if ($cacheEntry instanceof CacheEntry && $this->getMountOption('filesystem_check_changes', 1) !== 1) {
+				if ($cacheEntry instanceof CacheEntry) {
 					$stat['size'] = $cacheEntry->getSize();
 					$stat['mtime'] = $cacheEntry->getMTime();
+				} else {
+					//folders don't really exist
+					$stat['size'] = -1; //unknown
+					$stat['mtime'] = time();
```
@juliushaertl Could this be a solution ? Default to using cache and return dummy values if there is no cache.
It is working and I think it makes sense and feels cleaner. But maybe your insight would favor the `storage_mtime` solution.

artonge replied:
After discussion, this is the chosen solution.

Tests done:

- The size is correctly updated when navigating into folders
- The size is not reset to -1 on refresh

Notes:

- The `filesystem_check_changes` was removed as it seems useless here:
    - when it is set to `never`, it is safe to return the cached size as it reflects the displayed file structure
    - when it is set the `always`, the displayed file structure changes, but the cached size is correctly updated during navigation
 
Other fix tried:

- Preventing cache update if the size is -1 in `Scanner.php`. But the cache is updated by another code path, and some code paths might legitly want to set the size to -1.

artonge replied:
@juliushaertl if this ^ is ok with you, then let merge this

juliushaertl replied:
The scenario that won't work this this is if the files are changed or added directly on the external storage.

- Create a folder
- Put an object directly on the s3 bucket in the virtual path of the folder
- Access the folder

With filesystem_check_changes=always the folder should get scanned again due to the size -1 and updated mtime and the file should appear in Nextcloud.

juliushaertl replied:
Unfortunately just triggering the mtime update with something like 

```
if ($this->getMountOption('filesystem_check_changes', 1) === 1) {
	$stat['mtime'] = time();
}
```

seems to have the same effect on the pending size update. Though updating storage_mtime also does not trigger the scanning the folder again.

PVince81 replied:
@artonge right, we forgot to test the following case with the fix:

1. set external storage to allow update detection (filesystem_check_changes=1)
2. create a folder "test" directly on S3
3. upload a file "test/file" directly on S3
4. navigate into that folder with NC => the size updates, etc
5. upload another version of the same file name, with a different size
6. then navigate into "test" and "file" and see if the sizes updated

it is likely that the sizes did not update any more for this case

juliushaertl replied:
Thinking about that a bit more adding a dedicated hasUpdated method to the AmazonS3 implementation could help that just always returns true in addition to your changes, which would always trigger the updater then even if the size was not set to -1. 

```
diff --git a/apps/files_external/lib/Lib/Storage/AmazonS3.php b/apps/files_external/lib/Lib/Storage/AmazonS3.php
index bf5a988704..d560488a16 100644
--- a/apps/files_external/lib/Lib/Storage/AmazonS3.php
+++ b/apps/files_external/lib/Lib/Storage/AmazonS3.php
@@ -406,6 +406,10 @@ class AmazonS3 extends \OC\Files\Storage\Common {
                }
        }
 
+       public function hasUpdated($path, $time) {
+               return $this->getMountOption('filesystem_check_changes', 1) === 1 || parent::hasUpdated($path, $time);
+       }
+
        /**
         * Return content length for object
         *
```

CarlSchwan commented:
I investigated a similar issue with using Group Folders with the S3 storage with Vincent last week:

* OC\File\Cache\Updater: the update function is never called for the folder itself, only for the preview. It's correctly called for folders outside the group folder
* A S3ObjectStorage uses a NoopScanner, but the GroupFolderStorage uses a normal Scanner. This will probably cause troubles when the Updater::update method is called. Maybe it's the reason why Update is not called but I doubt it.

PVince81 commented:
ref: https://github.com/nextcloud/groupfolders/pull/1630

but probably not related because the latter is for group folders and the problem with S3 primary happens also without

artonge committed:
Fix folder size contained in S3 buckets

If 'filesystem_check_changes' was set to never, the cached size was alway set to -1 (Pending) on every access

Signed-off-by: Louis Chemineau <louis@chmn.me>

artonge commented:
/backport to stable22

artonge commented:
/backport to stable21

PVince81 reviewed (approved):
Alright, since the solution covers all the interesting cases let's move forward :+1:
